datetime.go

```diff
@@ -0,0 +1,233 @@
+package dur
+
+import (
+	"errors"
+	"strconv"
+	"strings"
+	"time"
+)
+
+//var absoluteTimeFormats = []string{"15:04 20060102", "20060102", "01/02/06"}
+
+var absoluteTimeFormats = []string{"15:04 20060102", "20060102", "01/02/06", "01/02/2006"}
+var errUnknownDateTimeFormat = errors.New("parse error. unknown DateTime format")
+var errUnknownTimeFormat = errors.New("parse error. unknown Time format")
+
+// ParseDateTime parses a format string to a unix timestamp, or error otherwise.
+// 'loc' is the timezone to use for interpretation (when applicable)
+// 'now' is a reference, in case a relative specification is given.
+// 'def' is a default in case an empty specification is given.
+func ParseDateTime(s string, loc *time.Location, now time.Time, def uint32) (uint32, error) {
+	now = now.In(loc)
+	switch s {
+	case "":
+		return uint32(def), nil
+	case "now":
+		return uint32(now.Unix()), nil
+	case "today":
+		year, month, day := now.Date()
+		return uint32(time.Date(year, month, day, 0, 0, 0, 0, loc).Unix()), nil
+	case "midnight":
+		year, month, day := now.Date()
+		return uint32(time.Date(year, month, day, 0, 0, 0, 0, loc).Unix()), nil
+	case "noon":
+		year, month, day := now.Date()
+		return uint32(time.Date(year, month, day, 12, 0, 0, 0, loc).Unix()), nil
+	case "teatime":
+		year, month, day := now.Date()
+		return uint32(time.Date(year, month, day, 16, 0, 0, 0, loc).Unix()), nil
+	case "yesterday":
+		base := now.AddDate(0, 0, -1)
+		year, month, day := base.Date()
+		return uint32(time.Date(year, month, day, 0, 0, 0, 0, loc).Unix()), nil
+	case "tomorrow":
+		base := now.AddDate(0, 0, 1)
+		year, month, day := base.Date()
+		return uint32(time.Date(year, month, day, 0, 0, 0, 0, loc).Unix()), nil
+	}
+
+	// try negative relative duration offset
+	if s[0] == '-' {
+		dur, err := ParseNDuration(s[1:])
+		if err == nil {
+			return uint32(now.Add(-time.Duration(dur) * time.Second).Unix()), nil
+		}
+		return 0, err
+	}
+
+	// not documented, though supported by graphite so we must do the same.
+	// https://github.com/raintank/metrictank/issues/673
+	if strings.HasPrefix(s, "now-") {
+		dur, err := ParseNDuration(s[4:])
+		if err == nil {
+			return uint32(now.Add(-time.Duration(dur) * time.Second).Unix()), nil
+		}
+		return 0, err
+	}
+	if strings.HasPrefix(s, "now+") {
+		dur, err := ParseNDuration(s[4:])
+		if err == nil {
+			return uint32(now.Add(time.Duration(dur) * time.Second).Unix()), nil
+		}
+		return 0, err
+	}
```

Comment:
graphite supports + and - syntax on any valid time specification, so this should probably be re-organized to take a similar approach by first trying to split out the modifier, then parsing the time specification, then applying the modifier.